### PR TITLE
stm32l4-multi: Moved uart to multilib

### PR DESCRIPTION
**Pull request by agkaminski**
WIP: In the next step I will move SPI driver to lib. **Please don't merge yet**, if this passes code review I will add SPI before merging.

### CODE REVIEW

File: multi/stm32l4-multi/flash.h [hunk 1]
```diff
@@ -15,14 +15,34 @@
...
+#ifndef FLASH_PROGRAM_1_ADDR
+#define FLASH_PROGRAM_1_ADDR 0x08000000
+#endif
+
+#ifndef FLASH_PROGRAM_2_ADDR
+#define FLASH_PROGRAM_2_ADDR 0x08080000
+#endif
+
+#ifndef FLASH_PROGRAM_BANK_SIZE
+#define FLASH_PROGRAM_BANK_SIZE (320 * 1024)
```
Comment (anglov) [inline multi/stm32l4-multi/flash.h:33]: I know that it's not a topic of this change but you sure that's valid value? 

Shouldn't be 512?

File: multi/stm32l4-multi/uart.c [hunk 2]
```diff
@@ -66,308 +46,42 @@ static const int uartConfig[] = { UART1, UART2, UART3, UART4, UART5, LPUART1 };
...
-		{ (void *)0x40013800, pctl_usart1, usart1_irq },
-		{ (void *)0x40004400, pctl_usart2, usart2_irq },
-		{ (void *)0x40004800, pctl_usart3, usart3_irq },
-		{ (void *)0x40004c00, pctl_uart4, uart4_irq },
-		{ (void *)0x40005000, pctl_uart5, uart5_irq },
-		{ (void *)0x40008000, pctl_lpuart1, lpuart1_irq }
-	};
 
-	for (i = 0, uart = 0; uart < sizeof(info) / sizeof(info[0]); ++uart) {
+	for (i = 0, uart = usart1; uart < uart5; ++uart) {
```
Comment (anglov) [inline multi/stm32l4-multi/uart.c:80]: You missed initialisation of uart5 this way.

File: multi/stm32l4-multi/lib/libuart.c [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+		volatile uint32_t *base;
+		int dev;
+		unsigned irq;
+	} info[] = {
+		{ (void *)0x40013800, pctl_usart1, usart1_irq },
+		{ (void *)0x40004400, pctl_usart2, usart2_irq },
+		{ (void *)0x40004800, pctl_usart3, usart3_irq },
+		{ (void *)0x40004c00, pctl_uart4, uart4_irq },
+		{ (void *)0x40005000, pctl_uart5, uart5_irq },
+		{ (void *)0x40008000, pctl_lpuart1, lpuart1_irq }
```
Comment (anglov) [inline multi/stm32l4-multi/lib/libuart.c:273]: Just want to be sure if it's your intention to remove reading/writing to lpuart from multidrv but initializing it in libuart anyway?

File: multi/stm32l4-multi/uart.c [hunk 4]
```diff
@@ -35,29 +37,7 @@
 
 #define UART_CNT (UART1 + UART2 + UART3 + UART4 + UART5 + LPUART1)
```
Comment (anglov) [inline multi/stm32l4-multi/uart.c:38]: If LPUART is removed from reading/writing here, shouldn't be removed from this macros?

**Review by anglov — CHANGES_REQUESTED** [review]
Looks good, only mistake I see is initialisation of uart5 and not sure about lpuart.

File: multi/stm32l4-multi/lib/libuart.c [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+	};
+
+	if (uart < 0 || uart > (int)(sizeof(info) / sizeof(info[0])))
```
Comment (kemonats) [inline multi/stm32l4-multi/lib/libuart.c:276]: Probably bad check of the array index.
What if:
`uart = sizeof(info) / sizeof(info[0])`
Should:
```
	if (uart < 0 || uart >= (int)(sizeof(info) / sizeof(info[0])))
```
Maybe it's better use `uart` as `unsigned int` for the array index
```
int libuart_init(libuart_ctx *ctx, unsigned int uart)
```
and
```
	if (uart >= sizeof(info) / sizeof(info[0]))
```

File: multi/stm32l4-multi/flash.h (hunk 1 above)
Comment (agkaminski) [reply multi/stm32l4-multi/flash.h]: Yes, fixed

File: multi/stm32l4-multi/uart.c (hunk 2 above)
Comment (agkaminski) [reply multi/stm32l4-multi/uart.c]: Fixed

File: multi/stm32l4-multi/lib/libuart.c (hunk 3 above)
Comment (agkaminski) [reply multi/stm32l4-multi/lib/libuart.c]: Yeah, missed that, removed

File: multi/stm32l4-multi/uart.c (hunk 4 above)
Comment (agkaminski) [reply multi/stm32l4-multi/uart.c]: It should be, fixed

File: multi/stm32l4-multi/lib/libuart.c (hunk 5 above)
Comment (agkaminski) [reply multi/stm32l4-multi/lib/libuart.c]: You're right, fixed that, that you

**Review by anglov — APPROVED** [review]

File: multi/stm32l4-multi/lib/libuart.c [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	ctx->base = info[uart].base;
+
+	ctx->txbeg = NULL;
+	ctx->txend = NULL;
+
+	ctx->rxbeg = NULL;
+	ctx->rxend = NULL;
+	ctx->read = NULL;
+	ctx->rxdr = 0;
+	ctx->rxdw = 0;
```
Comment (kemonats) [inline multi/stm32l4-multi/lib/libuart.c:296]: Initalization is in `libuart_configure()`

Comment (agkaminski) [reply multi/stm32l4-multi/lib/libuart.c]: Fixed, thanks

**Review by Darchiv — APPROVED** [review]

**Pull request merged**